client/pkg/infra/provision/context.go

```diff
@@ -0,0 +1,194 @@
+func WithoutConnectionParams() SchematicOption {
+	return func(so *SchematicOptions) {
+		so.skipConnectionParams = true
+	}
+}
+
+// WithExtraExtensions adds more extensions to the schematic.
+// The provider can detect the hardware and install some extensions automatically using this method.
+func WithExtraExtensions(extensions ...string) SchematicOption {
+	return func(so *SchematicOptions) {
+		so.extensions = extensions
+	}
+}
+
+// WithMetaValues adds meta values to the generated schematic.
+// If the meta values with the same names are already set they are overwritten.
+func WithMetaValues(values ...schematic.MetaValue) SchematicOption {
+	return func(so *SchematicOptions) {
+		so.metaValues = values
+	}
+}
+
+// WithExtraKernelArgs adds kernel args to the schematic.
+// This method doesn't remove duplicate kernel arguments.
+func WithExtraKernelArgs(args ...string) SchematicOption {
+	return func(so *SchematicOptions) {
+		so.kernelArgs = args
+	}
+}
+
+// WithOverlay sets the overlay on the schematic.
+func WithOverlay(overlay schematic.Overlay) SchematicOption {
+	return func(so *SchematicOptions) {
+		so.overlay = &overlay
+	}
+}
+
+// NewContext creates a new provision context.
+func NewContext[T resource.Resource](
+	machineRequest *infra.MachineRequest,
+	machineRequestStatus *infra.MachineRequestStatus,
+	state T,
+	connectionParams string,
+	imageFactory FactoryClient,
+) Context[T] {
+	return Context[T]{
+		machineRequest:       machineRequest,
+		MachineRequestStatus: machineRequestStatus,
+		State:                state,
+		ConnectionParams:     connectionParams,
+		imageFactory:         imageFactory,
+	}
+}
+
+// Context keeps all context which might be required for the provision calls.
+type Context[T resource.Resource] struct {
+	machineRequest       *infra.MachineRequest
+	imageFactory         FactoryClient
+	MachineRequestStatus *infra.MachineRequestStatus
+	State                T
+	ConnectionParams     string
+}
+
+// GetRequestID returns machine request id.
+func (context *Context[T]) GetRequestID() string {
+	return context.machineRequest.Metadata().ID()
+}
+
+// GetTalosVersion returns Talos version from the machine request.
+func (context *Context[T]) GetTalosVersion() string {
+	return context.machineRequest.TypedSpec().Value.TalosVersion
+}
+
+// SetMachineUUID in the machine request status.
+func (context *Context[T]) SetMachineUUID(value string) {
+	context.MachineRequestStatus.TypedSpec().Value.Id = value
+}
+
+// SetMachineInfraID in the machine request status.
+func (context *Context[T]) SetMachineInfraID(value string) {
+	context.MachineRequestStatus.Metadata().Labels().Set(omni.LabelMachineInfraID, value)
+}
+
+// GenerateSchematicID generate the final schematic out of the machine request.
+// This method also calls the image factory and uploads the schematic there.
+func (context *Context[T]) GenerateSchematicID(ctx context.Context, opts ...SchematicOption) (string, error) {
```

Comment:
being able to pass a logger here could be helpful by `.WithLogger` or something, and maybe we could log the generated schematic yaml

Reply:
makes sense